### PR TITLE
fix: JSON unmarshaling error handling and nil return

### LINKED ISSUE #18604: Health Lua script cannot return nil status object
**Describe the bug**

The goal is to only set the health on a resource when some conditions are, and not set any health on the object if the Lua script does not report any status.

```
  resource.customizations.health.Secret: |
    hs = {}

    if obj.metadata.annotations ~= nil and obj.metadata.annotations["my-annotation"] ~= nil and obj.metadata.annotations["my-annotation"] == "foo" then
      hs.status = "Healthy"
      hs.message = ""
      return hs
    end

    if obj.metadata.annotations ~= nil and obj.metadata.annotations["my-annotation"] ~= nil and obj.metadata.annotations["my-annotation"] == "bar" then
      hs.status = "Healthy"
      hs.message = ""
      return hs
    end

    if obj.metadata.annotations ~= nil and obj.metadata.annotations["my-annotation"] ~= nil then
      hs.status = "Degraded"
      hs.message = "this is an example"
      return hs
    end

    -- Here we want to return nil or empty object so that the health is not set
    return hs
```

When a health check on a resource is returning empty or nil from a Lua script, Argo is not able to ignore the Health of that resource.

**To Reproduce**

Setting the following script in argo-cm

```
  resource.customizations.health.Secret: |
    hs = {} 
    return hs
```
then it will return ` json: cannot unmarshal array into Go value of type health.HealthStatus`

If we set it instead to the following
```
  resource.customizations.health.Secret: |
    hs = {} 
    return nil
```
then it will return `expect table output from Lua script, not nil`

If we set it instead to the following
```
  resource.customizations.health.Secret: |
	-- A script that does not return anything
```
then it will return `expect table output from Lua script, not nil`

**Expected behavior**

`hs = {}` should be unmarshaled correctly and not set (and remove) any health status in ArgoCD
`return nil` should be valid and not set (and remove) any health status in ArgoCD
`return` should be valid and not set (and remove) any health status in ArgoCD

**Screenshots**

image

**Version**

2.11.2

**Workaround**

We can set the health status to healthy
```
  resource.customizations.health.Secret: |
    hs = {} 
    hs.status = "Healthy"
    return nil
```

### PR DESCRIPTION
**Pull request by Yuni-sa**
Closes #18604
Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [X] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [X] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [X] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [X] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [X] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: util/lua/lua.go [hunk 1]
```diff
@@ -109,6 +111,10 @@ func (vm VM) ExecuteHealthLua(obj *unstructured.Unstructured, script string) (*h
 		healthStatus := &health.HealthStatus{}
 		err = json.Unmarshal(jsonBytes, healthStatus)
 		if err != nil {
+			typeError := &json.UnmarshalTypeError{Value: "array", Type: reflect.TypeOf(healthStatus)}
+			if errors.As(err, &typeError) {
+				return &health.HealthStatus{}, nil
+			}
```
Comment (agaudreault) [inline util/lua/lua.go:117]: Let's add a comment as this is non-trivial to know 

```suggestion
			// Validate if the error is caused by an empty object
			typeError := &json.UnmarshalTypeError{Value: "array", Type: reflect.TypeOf(healthStatus)}
			if errors.As(err, &typeError) {
				return &health.HealthStatus{}, nil
			}
```

**Review by agaudreault — APPROVED** [review]
LGTM 👍

File: util/lua/lua.go (hunk 1 above)
Comment (Yuni-sa) [reply util/lua/lua.go:118]: good idea, Thanks!

**Review by ishitasequeira — APPROVED** [review]
LGTM!!

**Pull request merged**